rosetta/validator/block.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2021 Alvalor S.A.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package validator
+
+import (
+	"fmt"
+
+	"github.com/onflow/flow-go/model/flow"
+
+	"github.com/optakt/flow-dps/rosetta/failure"
+	"github.com/optakt/flow-dps/rosetta/identifier"
+)
+
+// Block identifier tries to extrapolate the block identifier to a full version
+// of itself. For now, we will always need a height.
+// NOTE: We always pass a block identifier that in principle at least could be
+// valid, so we will have at least a height or a hash.
+func (v *Validator) Block(block identifier.Block) (identifier.Block, error) {
+
+	// We currently only support retrieval by height, until we start indexing
+	// the block IDs as part of the DPS index.
+	if block.Index == 0 {
+		return identifier.Block{}, fmt.Errorf("block access with hash currently not supported")
+	}
+
+	// We should always be able to parse this at this point, if it is present,
+	// as we already checked the format, so normal error is fine.
+	var blockID flow.Identifier
+	var err error
+	if block.Hash != "" {
+		blockID, err = flow.HexStringToIdentifier(block.Hash)
+		if err != nil {
+			return identifier.Block{}, failure.InvalidBlock{
+				Height:  block.Index,
+				BlockID: flow.ZeroID,
+				Message: fmt.Sprintf("block hash not in hex format (hash: %s)", block.Hash),
+			}
+		}
+	}
+
+	// The block index can't be below the first indexed height.
+	first, err := v.index.First()
+	if err != nil {
+		return identifier.Block{}, fmt.Errorf("could not get first: %w", err)
+	}
```

Comment:
Wouldn't it make sense to store the first indexed height in the validator, to avoid repeated writes on the index and remove one instance of error handling from this func?

Reply:
We can't really make assumptions about the first height, to be honest. While the current Rosetta API is exposed in front of a single DPS, with a static first height, future versions might be exposed in front of a multitude of spork DPS servers, and we might add them at the back as we grow our past state access.